consensus/view_change.go

```diff
@@ -224,88 +265,34 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
 	}
 
 	// received enough view change messages, change state to normal consensus
-	if consensus.Decider.IsQuorumAchievedByMask(consensus.vc.GetViewIDBitmap(recvMsg.ViewID)) {
-		consensus.getLogger().Info().Msg("[onViewChange] View Change Message Quorum Reached")
-		consensus.current.SetMode(Normal)
-		consensus.LeaderPubKey = newLeaderKey
-		consensus.ResetState()
+	if consensus.Decider.IsQuorumAchievedByMask(consensus.vc.GetViewIDBitmap(recvMsg.ViewID)) && consensus.IsViewChangingMode() {
+		// no previous prepared message, go straight to normal mode
+		// and start proposing new block
 		if consensus.vc.IsM1PayloadEmpty() {
-			// TODO(Chao): explain why ReadySignal is sent only in this case but not the other case.
-			// Make sure the newly proposed block have the correct view ID
-			consensus.SetCurBlockViewID(recvMsg.ViewID)
-			go func() {
-				consensus.ReadySignal <- struct{}{}
-			}()
-		} else {
-			consensus.switchPhase("onViewChange", FBFTCommit)
-			payload := consensus.vc.GetM1Payload()
-			copy(consensus.blockHash[:], payload[:32])
-			aggSig, mask, err := consensus.ReadSignatureBitmapPayload(payload, 32)
-
-			if err != nil {
-				consensus.getLogger().Error().Err(err).
-					Msg("[onViewChange] ReadSignatureBitmapPayload Fail")
-				return
-			}
-
-			consensus.aggregatedPrepareSig = aggSig
-			consensus.prepareBitmap = mask
-			// Leader sign and add commit message
-			block := consensus.FBFTLog.GetBlockByHash(consensus.blockHash)
-			if block == nil {
-				consensus.getLogger().Warn().Msg("[onViewChange] failed to get prepared block for self commit")
+			if err := consensus.startNewView(recvMsg.ViewID, newLeaderPriKey); err != nil {
+				consensus.getLogger().Error().Err(err).Msg("[onViewChange] startNewView failed")
 				return
 			}
-			commitPayload := signature.ConstructCommitPayload(consensus.ChainReader,
-				block.Epoch(), block.Hash(), block.NumberU64(), block.Header().ViewID().Uint64())
-			for i, key := range consensus.priKey {
-				if err := consensus.commitBitmap.SetKey(key.Pub.Bytes, true); err != nil {
-					consensus.getLogger().Warn().Err(err).
-						Msgf("[OnViewChange] New Leader commit bitmap set failed for key at index %d", i)
-					continue
-				}
-
-				if _, err := consensus.Decider.SubmitVote(
-					quorum.Commit,
-					key.Pub.Bytes,
-					key.Pri.SignHash(commitPayload),
-					common.BytesToHash(consensus.blockHash[:]),
-					block.NumberU64(),
-					block.Header().ViewID().Uint64(),
-				); err != nil {
-					consensus.getLogger().Warn().Err(err).Msg("submit vote on viewchange commit failed")
-					return
-				}
+			consensus.ResetState()
+			consensus.LeaderPubKey = newLeaderKey
```

Comment:
why not put 
consensus.ResetState()
consensus.LeaderPubKey = newLeaderKey

inside startNewView?   they need to be lock protected too.

Reply:
make sense.